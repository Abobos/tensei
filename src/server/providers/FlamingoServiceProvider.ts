import Fs from 'fs'
import Path from 'path'
import Dotenv from 'dotenv'
import Express from 'express'
import Mongodb from 'mongodb'
import BodyParser from 'body-parser'
import ConnectMongo from 'connect-mongo'
import ExpressSession from 'express-session'
import {
    Config,
    Resource,
    FlamingoServiceProviderInterface,
} from './../typings/interfaces'
import ClientController from '../controllers/ClientController'
import LoginController from '../controllers/auth/LoginController'

import DatabaseRepository from '../database/Repository'

class FlamingoServiceProvider implements FlamingoServiceProviderInterface {
    public app: Express.Application = Express()

    public client: Mongodb.MongoClient | null = null

    public router: Express.Router = Express.Router()

    public resources: Resource[] = []

    public db: DatabaseRepository | null = null

    public config: Config | null = null

    constructor(public readonly $root: string) {}
    /**
     *
     * Get the path to where the resources are saved.
     *
     * @param root this is the root of the project
     */
    public resourcesIn(root: string) {
        return Path.resolve(root, 'resources')
    }

    public async register() {
        this.registerEnvironmentVariables()

        await this.registerResources()

        this.db = await this.establishDatabaseConnection()

        this.registerMiddleware()

        await this.registerRoutes()
    }

    public registerEnvironmentVariables() {
        Dotenv.config({
            path: Path.resolve(this.$root, '.env'),
        })

        this.registerConfig()
    }

    public registerMiddleware() {
        this.app.use(BodyParser.json())

        const Store = ConnectMongo(ExpressSession)

        this.app.use(
            (
                request: Express.Request,
                response: Express.Response,
                next: Express.NextFunction
            ) => {
                request.db = this.db
                request.resources = this.resources

                next()
            }
        )

        this.app.use(
            ExpressSession({
                secret: this.config?.sessionSecret!,
                store: new Store({
                    client: this.client!,
                }),
                resave: false,
                saveUninitialized: false,
            })
        )
    }

    public async registerRoutes() {
        this.registerAuthRoutes()

        this.router.get('*', ClientController.index)

        this.app.use(this.router)
    }

    public launchServer(serverCallback: (config: Config) => void) {
        this.app.listen(this.config!.port, () => serverCallback(this.config!))
    }

    public registerAuthRoutes() {
        this.router.post('/api/login', LoginController.store)
    }

    public async registerResources() {
        const resourcesPath = this.resourcesIn(this.$root)

        this.resources = Fs.readdirSync(resourcesPath)

            .filter((file) => file.substring(file.length - 3) === '.js')

            .map((file) => {
                const Resource = require(Path.resolve(resourcesPath, file))

                return Resource.default
                    ? new Resource.default()
                    : new Resource()
            })
    }

    public async establishDatabaseConnection() {
        this.client = new Mongodb.MongoClient(
            process.env.DATABASE_URI as string,
            {
                useNewUrlParser: true,
                useUnifiedTopology: true,
            }
        )

        await this.client.connect()

        return new DatabaseRepository(this.client.db())
    }

    private registerConfig() {
        this.config = {
            port: process.env.PORT || 1377,
            sessionSecret: process.env.SESSION_SECRET || 'test-session-secret',
            databaseUri:
                process.env.DATABASE_URI || 'mongodb://localhost/flamingo',
        }
    }
}

export default FlamingoServiceProvider
