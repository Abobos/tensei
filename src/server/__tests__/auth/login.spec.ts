import Path from 'path'
import Bcrypt from 'bcryptjs'
import Supertest from 'supertest'
import FlamingoServiceProvider from '../../providers/FlamingoServiceProvider'

describe('The login endpoint', () => {
    process.env.DATABASE_URI = 'mongodb://localhost/flamingo-testdb'

    const setup = async () => {
        let instance = new FlamingoServiceProvider(
        )

        await instance.register()

        return instance
    }

    it('validates the data before processing the login', async () => {
        const { app, client: dbClient } = await setup()

        const client = Supertest(app)

        const response = await client.post('/api/login').send({})

        expect(response.status).toBe(422)
        expect(response.body).toMatchSnapshot()

        await dbClient?.close()
    })

    it('returns validation errors if user is not found', async () => {
        const { app, client: dbClient } = await setup()

        const client = Supertest(app)

        const response = await client.post('/api/login').send({
            email: 'hey@unknown-user.io',
            password: 'password',
        })

        expect(response.status).toBe(422)
        expect(response.body).toMatchSnapshot()

        await dbClient?.close()
    })

    it('can correctly login an administrator', async () => {
        const instance = await setup()

        await instance.db?.admin().deleteMany({})

        const testFirstName = 'Hey'
        const testPassword = 'password'
        const testLastName = 'Flamingo'
        const testEmail = 'hey@flamingo.io'

        const client = Supertest(instance.app)

        await instance.db?.admin().insertOne({
            email: testEmail,
            password: Bcrypt.hashSync(testPassword),
            firstName: testFirstName,
            lastName: testLastName,
        })

        const response = await client.post('/api/login').send({
            email: testEmail,
            password: testPassword,
        })

        expect(response.status).toBe(200)

        await instance.client?.close()
    })
})
