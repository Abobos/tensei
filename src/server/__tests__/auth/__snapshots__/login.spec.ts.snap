// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The login endpoint returns validation errors if user is not found 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "These credentials do not match our records.",
    },
  ],
  "message": "Validation failed.",
}
`;

exports[`The login endpoint validates the data before processing the login 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "The email is required.",
      "validation": "required",
    },
    Object {
      "field": "password",
      "message": "The password is required.",
      "validation": "required",
    },
  ],
  "message": "Validation failed.",
}
`;
